server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java

```diff
@@ -172,6 +176,73 @@ public void collect(int doc, long bucket) throws IOException {
         directory.close();
     }
 
+    /**
+     * Check that knn queries rewrite parallelizes on the number of segments if there are enough
+     * threads available.
+     */
+    public void testConcurrentKnnRewrite() throws Exception {
+        final Directory directory = newDirectory();
+        try (
+            IndexWriter iw = new IndexWriter(
+                directory,
+                new IndexWriterConfig(new StandardAnalyzer()).setMergePolicy(NoMergePolicy.INSTANCE)
+            )
+        ) {
+            final int numDocs = randomIntBetween(100, 200);
+            for (int i = 0; i < numDocs; i++) {
+                Document document = new Document();
+                document.add(new StringField("field", "value", Field.Store.NO));
+                iw.addDocument(document);
+                if (rarely()) {
+                    iw.commit();
+                }
+            }
+        }
+
+        DirectoryReader directoryReader = DirectoryReader.open(directory);
+        int numSegments = directoryReader.getContext().leaves().size();
+        // make sure we have more threads than segments available to check later call to execute method
+        int nThreads = randomIntBetween(numSegments, numSegments + 5);
```

Comment:
if you have an unbounded queue, like I think you have, the number of threads does not seem to matter for the assertions you have. tasks will queue up and be executed by less threads, but the number of execute calls will always be the same as the number of tasks. Rejections won't happen because the queue is unbounded.

Reply:
This is contrary to what I saw earlier when I was writing this test. It seems to be controlled by the [executors queue size](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java#L578). While the capacity of the `LinkedBlockingQueue` is unbounded, the actual queue size is not. I can set the maximumPoolSize of the executor to a higher value though, since this seems to be taken into account here as well.